## Data Understanding for Customer Segmentation and Retention Analysis

In this notebook, we will perform data understanding to explore the dataset, analyze customer behavior, and identify key patterns related to segmentation and retention.

In [21]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt


## Load Dataset

In [17]:
df = pd.read_csv("E:\\Netflix Customer Segmentation & Retention Analysis\\Data\\netflix_user_behavior_churn_50000v2.csv")

In [18]:
df.head()

,user_id,age,gender,region,subscription_type,payment_method,primary_device,account_age_months,favorite_genre,time_of_day,recommendation_source,session_count,avg_watch_time_minutes_per_week,watch_sessions_per_week,completion_rate,avg_rating_given,app_rating,recommendation_click_rate,days_since_last_login,churned
0,N00001,33,Male,Europe,Basic,Credit Card,Smart TV,38,Comedy,Afternoon,Trending,1,259,4,79,4,4,48,3,0
1,N00002,41,Male,Africa,Premium,Credit Card,Tablet,49,Comedy,Evening,Search,1,164,4,69,4,4,32,30,0
2,N00003,28,Female,Europe,Standard,Gift Card,Laptop,20,Drama,Evening,Ads,1,304,5,75,4,3,31,5,1
3,N00004,40,Female,North America,Basic,Credit Card,Mobile,19,Action,Evening,Algorithm,1,145,6,79,4,4,39,38,0
4,N00005,54,Male,Europe,Standard,Paypal,Mobile,51,Romance,Evening,Friend,3,108,1,58,5,4,50,5,0


## Data Overview

In [22]:
df.shape

(50000, 20)

In [19]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column                           Non-Null Count  Dtype
---  ------                           --------------  -----
 0   user_id                          50000 non-null  str  
 1   age                              50000 non-null  int64
 2   gender                           50000 non-null  str  
 3   region                           50000 non-null  str  
 4   subscription_type                50000 non-null  str  
 5   payment_method                   50000 non-null  str  
 6   primary_device                   50000 non-null  str  
 7   account_age_months               50000 non-null  int64
 8   favorite_genre                   50000 non-null  str  
 9   time_of_day                      50000 non-null  str  
 10  recommendation_source            50000 non-null  str  
 11  session_count                    50000 non-null  int64
 12  avg_watch_time_minutes_per_week  50000 non-null  int64
 1

In [20]:
df.describe()

,age,account_age_months,session_count,avg_watch_time_minutes_per_week,watch_sessions_per_week,completion_rate,avg_rating_given,app_rating,recommendation_click_rate,days_since_last_login,churned
count,50000.000000,50000.000000,50000.000000,50000.000000,50000.000000,50000.000000,50000.000000,50000.000000,50000.000000,50000.000000,50000.000000
mean,39.215900,44.803340,1.630120,277.950640,5.731460,73.127260,3.844560,3.946220,38.640680,12.301380,0.209280
std,13.071976,21.890139,0.988579,157.604512,3.467176,11.148734,0.611887,0.639168,19.474178,15.453764,0.406799
min,19.000000,1.000000,1.000000,15.000000,1.000000,23.000000,2.000000,2.000000,0.000000,0.000000,0.000000
25%,31.000000,29.000000,1.000000,165.000000,3.000000,66.000000,3.000000,4.000000,24.000000,1.000000,0.000000
50%,37.000000,45.000000,1.000000,240.000000,5.000000,74.000000,4.000000,4.000000,37.000000,4.000000,0.000000
75%,45.000000,60.000000,2.000000,349.000000,7.000000,81.000000,4.000000,4.000000,52.000000,22.000000,0.000000
max,79.000000,135.000000,5.000000,820.000000,19.000000,100.000000,5.000000,5.000000,100.000000,60.000000,1.000000


## Target Variable Analysis

In [24]:
df['churned'].describe()

count    50000.000000
mean         0.209280
std          0.406799
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max          1.000000
Name: churned, dtype: float64

## Missing Value & Duplicate Row/Values Analysis

In [26]:
df.isnull().sum()

user_id                            0
age                                0
gender                             0
region                             0
subscription_type                  0
payment_method                     0
primary_device                     0
account_age_months                 0
favorite_genre                     0
time_of_day                        0
recommendation_source              0
session_count                      0
avg_watch_time_minutes_per_week    0
watch_sessions_per_week            0
completion_rate                    0
avg_rating_given                   0
app_rating                         0
recommendation_click_rate          0
days_since_last_login              0
churned                            0
dtype: int64

In [27]:
df.duplicated().sum()

np.int64(0)

## Identify numerical/categorical features

In [47]:
numerical_features = df.select_dtypes(include=['int64', 'float64']).columns.tolist()
categorical_features = df.select_dtypes(include=['object']).columns.tolist()

C:\Users\HP\AppData\Local\Temp\ipykernel_15528\150583915.py:2: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  categorical_features = df.select_dtypes(include=['object']).columns.tolist()


In [48]:
numerical_features

['age',
 'account_age_months',
 'session_count',
 'avg_watch_time_minutes_per_week',
 'watch_sessions_per_week',
 'completion_rate',
 'avg_rating_given',
 'app_rating',
 'recommendation_click_rate',
 'days_since_last_login',
 'churned']

In [49]:
categorical_features

['user_id',
 'gender',
 'region',
 'subscription_type',
 'payment_method',
 'primary_device',
 'favorite_genre',
 'time_of_day',
 'recommendation_source']

### Unique Value Analysis

In [40]:
df.nunique()

user_id                            50000
age                                   61
gender                                 2
region                                 6
subscription_type                      3
payment_method                         4
primary_device                         4
account_age_months                   131
favorite_genre                         7
time_of_day                            4
recommendation_source                  5
session_count                          5
avg_watch_time_minutes_per_week      801
watch_sessions_per_week               19
completion_rate                       74
avg_rating_given                       4
app_rating                             4
recommendation_click_rate            101
days_since_last_login                 61
churned                                2
dtype: int64

## Categorical Feature Analysis

In [53]:
df[categorical_features].value_counts().T


user_id  gender  region         subscription_type  payment_method  primary_device  favorite_genre  time_of_day  recommendation_source
N00001   Male    Europe         Basic              Credit Card     Smart TV        Comedy          Afternoon    Trending                 1
N00002   Male    Africa         Premium            Credit Card     Tablet          Comedy          Evening      Search                   1
N00003   Female  Europe         Standard           Gift Card       Laptop          Drama           Evening      Ads                      1
N00004   Female  North America  Basic              Credit Card     Mobile          Action          Evening      Algorithm                1
N00005   Male    Europe         Standard           Paypal          Mobile          Romance         Evening      Friend                   1
                                                                                                                                        ..
N49996   Female  Europe         

### Numerical Feature Analysis


In [50]:
#Removing the Target Variable
numerical_features.remove('churned')

In [51]:
df[numerical_features].describe().T


,count,mean,std,min,25%,50%,75%,max
age,50000.0,39.21590,13.071976,19.0,31.0,37.0,45.0,79.0
account_age_months,50000.0,44.80334,21.890139,1.0,29.0,45.0,60.0,135.0
session_count,50000.0,1.63012,0.988579,1.0,1.0,1.0,2.0,5.0
avg_watch_time_minutes_per_week,50000.0,277.95064,157.604512,15.0,165.0,240.0,349.0,820.0
watch_sessions_per_week,50000.0,5.73146,3.467176,1.0,3.0,5.0,7.0,19.0
completion_rate,50000.0,73.12726,11.148734,23.0,66.0,74.0,81.0,100.0
avg_rating_given,50000.0,3.84456,0.611887,2.0,3.0,4.0,4.0,5.0
app_rating,50000.0,3.94622,0.639168,2.0,4.0,4.0,4.0,5.0
recommendation_click_rate,50000.0,38.64068,19.474178,0.0,24.0,37.0,52.0,100.0
days_since_last_login,50000.0,12.30138,15.453764,0.0,1.0,4.0,22.0,60.0


## 🔍 Key Findings from Data Understanding

* The dataset contains **50,000 customer records and 20 features**.
* There are **no missing values** across any of the 20 columns.
* There are **no duplicate records**, indicating that each customer record is unique in the dataset.
* The dataset contains **11 numerical features** and **9 categorical features**.
* `user_id` is a unique identifier with **50,000 unique values** and will primarily be used for customer identification rather than predictive modeling.
* The numerical features cover **customer demographics, account tenure, engagement behavior, content interaction, ratings, recommendation interaction, and recency of activity**.
* The categorical features describe **gender, region, subscription type, payment method, primary device, favorite genre, time of day, and recommendation source**.
* The target variable, `churned`, is binary with two possible values: **0 (not churned)** and **1 (churned)**.
* The average churn rate is approximately **20.93%**, meaning around one in five customers in the dataset are labeled as churned.
* Customer ages range from **19 to 79 years**, with an average age of approximately **39 years**.
* Account tenure ranges from **1 to 135 months**, with a median of **45 months**.
* `avg_watch_time_minutes_per_week` ranges from **15 to 820 minutes**, with a median of **240 minutes per week**.
* `watch_sessions_per_week` ranges from **1 to 19 sessions**, with a median of **5 sessions per week**.
* `completion_rate` ranges from **23% to 100%**, with a median of **74%**.
* `recommendation_click_rate` ranges from **0% to 100%**, with a median of **37%**.
* `days_since_last_login` ranges from **0 to 60 days**, with a median of **4 days**, providing a useful measure of customer recency and potential inactivity.
* The categorical variables have relatively small numbers of unique categories, making them suitable for further categorical analysis and encoding during modeling.
* The dataset contains a combination of **demographic, subscription, engagement, behavioral, and recency-related features**, making it suitable for both **customer segmentation and churn prediction**.

### 🎯 Initial Understanding

The dataset provides sufficient information to investigate customer behavior from multiple perspectives. The next stages of the analysis will focus on **exploratory data analysis, relationships between customer characteristics and churn, behavioral segmentation, feature engineering, and predictive modeling**.
